web_monitoring/internetarchive.py

```diff
@@ -97,30 +184,151 @@ def cdx_hash(content):
+        subsequent retries increase by powers of 2:
+            seconds = backoff * 2 ^ (retry number - 1)
+        So if this was `4`, retries would happen after the following delays:
+            0 seconds, 4 seconds, 8 seconds, 16 seconds, ...
+    timeout : int or float or tuple of (int or float, int or float), optional
+        A timeout to use for all requests. If not set, there will be no
+        no explicit timeout. See the Requests docs for more:
+        http://docs.python-requests.org/en/master/user/advanced/#timeouts
+    user_agent : str, optional
+        A custom user-agent string to use in all requests.
+    """
 
+    # It seems Wayback sometimes produces 500 errors for transient issues, so
+    # they make sense to retry here. Usually not in other contexts, though.
+    retryable_statuses = frozenset((413, 421, 429, 500, 502, 503, 504, 599))
+
+    retryable_errors = (ConnectTimeoutError, MaxRetryError, ReadTimeoutError,
+                        ProxyError, RetryError, Timeout)
+    handleable_errors = (ConnectionError,) + retryable_errors
+
+    def __init__(self, retries=6, backoff=2, timeout=None, user_agent=None):
+        super().__init__()
+        self.retries = retries
+        self.backoff = backoff
+        self.timeout = timeout
+        self.headers = {
+            'User-Agent': user_agent or f'edgi.web_monitoring.WaybackClient/{__version__}',
+            'Accept-Encoding': 'gzip, deflate'
+        }
+        # NOTE: the nice way to accomplish retry/backoff is with a urllib3:
+        #     adapter = requests.adapters.HTTPAdapter(
+        #         max_retries=Retry(total=5, backoff_factor=2,
+        #                           status_forcelist=(503, 504)))
+        #     self.mount('http://', adapter)
+        # But Wayback mementos can have errors, which complicates things. See:
+        # https://github.com/urllib3/urllib3/issues/1445#issuecomment-422950868
+
+    # Customize the built-in `send` functionality with retryability.
+    # NOTE: worth considering whether we should push this logic to a custom
+    # requests.adapters.HTTPAdapter
     def send(self, *args, **kwargs):
-        if self._closed:
-            raise SessionClosedError('This session has already been closed '
-                                     'and cannot send new HTTP requests.')
-
-        return super().send(*args, **kwargs)
-
+        if self.timeout is not None and 'timeout' not in kwargs:
+            kwargs['timeout'] = self.timeout
 
-class WaybackClient:
+        total_time = 0
+        maximum = self.retries
+        retries = 0
+        while True:
+            try:
+                result = super().send(*args, **kwargs)
+                if retries >= maximum or not self.should_retry(result):
+                    return result
+            except WaybackSession.handleable_errors as error:
+                if retries >= maximum:
+                    raise WaybackRetryError(retries, total_time, error)
```

Comment:
To retain the full traceback
```suggestion
                    raise WaybackRetryError(retries, total_time, error) from error
```

Reply:
Ohh, cool, I did not know you could do that! 😲